### ISSUE TITLE
Compile error in nanoGPT

### ISSUE THREAD

**Issue by wjfwzzc** [pytorch/pytorch#95445]
### 🐛 Describe the bug

Use `pytorch==2.0.0dev20230213` or earlier, everything works fine. However after that (maybe due to the updated triton), the model in [nanoGPT](https://github.com/karpathy/nanoGPT) fails in the following code:

```python3
# git clone the repo or just download https://raw.githubusercontent.com/karpathy/nanoGPT/master/model.py

import torch
from model import GPT, GPTConfig

gpt = GPT(GPTConfig()).cuda()
gpt_compiled = torch.compile(gpt)
x = torch.ones((1, 1024), dtype=torch.int64).cuda()  # a fake input
gpt_compiled(x)  # throw errors below
```

### Error logs

```python3
error: block with no terminator, has llvm.cond_br %805, ^bb2, ^bb3
Pass execution failedLLVM ERROR: Failed to translate TritonGPU to LLVM IR.
---------------------------------------------------------------------------
BrokenProcessPool                         Traceback (most recent call last)
File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py:704, in OutputGraph.call_user_compiler(self, gm)
    703 else:
--> 704     compiled_fn = compiler_fn(gm, self.fake_example_inputs())
    705 _step_logger()(logging.INFO, f"done compiler function {name}")

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/debug_utils.py:1055, in wrap_backend_debug.<locals>.debug_wrapper(gm, example_inputs, **kwargs)
   1054 else:
-> 1055     compiled_gm = compiler_fn(gm, example_inputs)
   1057 return compiled_gm

File /usr/local/lib/python3.9/dist-packages/torch/__init__.py:1390, in _TorchCompileInductorWrapper.__call__(self, model_, inputs_)
   1388 from torch._inductor.compile_fx import compile_fx
-> 1390 return compile_fx(model_, inputs_, config_patches=self.config)

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/compile_fx.py:455, in compile_fx(model_, example_inputs_, inner_compile, config_patches)
    450 with overrides.patch_functions():
    451
    452     # TODO: can add logging before/after the call to create_aot_dispatcher_function
    453     # in torch._functorch/aot_autograd.py::aot_module_simplified::aot_function_simplified::new_func
    454     # once torchdynamo is merged into pytorch
--> 455     return aot_autograd(
    456         fw_compiler=fw_compiler,
    457         bw_compiler=bw_compiler,
    458         decompositions=select_decomp_table(),
    459         partition_fn=functools.partial(
    460             min_cut_rematerialization_partition, compiler="inductor"
    461         ),
    462         keep_inference_input_mutations=True,
    463     )(model_, example_inputs_)

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/backends/common.py:48, in aot_autograd.<locals>.compiler_fn(gm, example_inputs)
     47 with enable_aot_logging():
---> 48     cg = aot_module_simplified(gm, example_inputs, **kwargs)
     49     counters["aot_autograd"]["ok"] += 1

File /usr/local/lib/python3.9/dist-packages/torch/_functorch/aot_autograd.py:2798, in aot_module_simplified(mod, args, fw_compiler, bw_compiler, partition_fn, decompositions, hasher_type, static_argnums, keep_inference_input_mutations)
   2796 full_args.extend(args)
-> 2798 compiled_fn = create_aot_dispatcher_function(
   2799     functional_call,
   2800     full_args,
   2801     aot_config,
   2802 )
   2804 # TODO: There is something deeply wrong here; compiled_fn running with
   2805 # the boxed calling convention, but aot_module_simplified somehow
   2806 # historically returned a function that was not the boxed calling
   2807 # convention.  This should get fixed...

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/utils.py:163, in dynamo_timed.<locals>.dynamo_timed_inner.<locals>.time_wrapper(*args, **kwargs)
    162 t0 = time.time()
--> 163 r = func(*args, **kwargs)
    164 time_spent = time.time() - t0

File /usr/local/lib/python3.9/dist-packages/torch/_functorch/aot_autograd.py:2491, in create_aot_dispatcher_function(flat_fn, flat_args, aot_config)
   2489 # You can put more passes here
-> 2491 compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config)
   2493 if not hasattr(compiled_fn, "_boxed_call"):

File /usr/local/lib/python3.9/dist-packages/torch/_functorch/aot_autograd.py:1713, in aot_wrapper_dedupe(flat_fn, flat_args, aot_config, compiler_fn)
   1712     if ok:
-> 1713         return compiler_fn(flat_fn, leaf_flat_args, aot_config)
   1715 # Strategy 2: Duplicate specialize.
   1716 #
   1717 # In Haskell types, suppose you have:
   (...)
   1749 #   }
   1750 #   keep_arg_mask = [True, True, False, True]

File /usr/local/lib/python3.9/dist-packages/torch/_functorch/aot_autograd.py:2130, in aot_dispatch_autograd(flat_fn, flat_args, aot_config)
   2129     with track_graph_compiling(aot_config, "forward"):
-> 2130         compiled_fw_func = aot_config.fw_compiler(
   2131             fw_module, flat_args_with_views_handled
   2132         )
   2134 class CompiledFunction(torch.autograd.Function):

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/utils.py:163, in dynamo_timed.<locals>.dynamo_timed_inner.<locals>.time_wrapper(*args, **kwargs)
    162 t0 = time.time()
--> 163 r = func(*args, **kwargs)
    164 time_spent = time.time() - t0

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/compile_fx.py:430, in compile_fx.<locals>.fw_compiler(model, example_inputs)
    429 model = convert_outplace_to_inplace(model)
--> 430 return inner_compile(
    431     model,
    432     example_inputs,
    433     num_fixed=fixed,
    434     cudagraphs=cudagraphs,
    435     graph_id=graph_id,
    436 )

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/debug_utils.py:595, in wrap_compiler_debug.<locals>.debug_wrapper(gm, example_inputs, **kwargs)
    594 else:
--> 595     compiled_fn = compiler_fn(gm, example_inputs)
    597 return compiled_fn

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/debug.py:239, in DebugContext.wrap.<locals>.inner(*args, **kwargs)
    238 with DebugContext():
--> 239     return fn(*args, **kwargs)

File /usr/lib/python3.9/contextlib.py:79, in ContextDecorator.__call__.<locals>.inner(*args, **kwds)
     78 with self._recreate_cm():
---> 79     return func(*args, **kwds)

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/compile_fx.py:177, in compile_fx_inner(gm, example_inputs, cudagraphs, num_fixed, is_backward, graph_id)
    176         graph.run(*example_inputs)
--> 177         compiled_fn = graph.compile_to_fn()
    179 if cudagraphs:

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/graph.py:600, in GraphLowering.compile_to_fn(self)
    599 def compile_to_fn(self):
--> 600     return self.compile_to_module().call

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/utils.py:163, in dynamo_timed.<locals>.dynamo_timed_inner.<locals>.time_wrapper(*args, **kwargs)
    162 t0 = time.time()
--> 163 r = func(*args, **kwargs)
    164 time_spent = time.time() - t0

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/graph.py:589, in GraphLowering.compile_to_module(self)
    587     print(code)
--> 589 mod = PyCodeCache.load(code)
    590 for name, value in self.constants.items():

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/codecache.py:609, in PyCodeCache.load(cls, source_code, extra)
    608 mod.key = key
--> 609 exec(code, mod.__dict__, mod.__dict__)
    610 # another thread might set this first

File /tmp/torchinductor_root/o2/co2drangyxpn7bpt4cebwthtjwmxd6xidw5lzhnhyydb5rg4qfw7.py:569
    543 triton__10 = async_compile.triton('''
    544 import triton
    545 import triton.language as tl
   (...)
    565     tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp5, xmask)
    566 ''')
--> 569 async_compile.wait(globals())
    570 del async_compile

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/codecache.py:796, in AsyncCompile.wait(self, scope)
    795 if isinstance(result, (Future, TritonFuture)):
--> 796     scope[key] = result.result()
    797     pbar.update(1)

File /usr/local/lib/python3.9/dist-packages/torch/_inductor/codecache.py:654, in TritonFuture.result(self)
    653 # If the worker failed this will throw an exception.
--> 654 self.future.result()
    655 kernel = self.kernel = _load_kernel(self.source_code)

File /usr/lib/python3.9/concurrent/futures/_base.py:440, in Future.result(self, timeout)
    439 elif self._state == FINISHED:
--> 440     return self.__get_result()
    441 else:

File /usr/lib/python3.9/concurrent/futures/_base.py:389, in Future.__get_result(self)
    388 if self._exception:
--> 389     raise self._exception
    390 else:

BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.

The above exception was the direct cause of the following exception:

BackendCompilerFailed                     Traceback (most recent call last)
Cell In[6], line 1
----> 1 gpt_compiled(x)

File /usr/local/lib/python3.9/dist-packages/torch/nn/modules/module.py:1501, in Module._call_impl(self, *args, **kwargs)
   1496 # If we don't have any hooks, we want to skip the rest of the logic in
   1497 # this function, and just call forward.
   1498 if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
   1499         or _global_backward_pre_hooks or _global_backward_hooks
   1500         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501     return forward_call(*args, **kwargs)
   1502 # Do not call functions when jit is used
   1503 full_backward_hooks, non_full_backward_hooks = [], []

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/eval_frame.py:82, in OptimizedModule.forward(self, *args, **kwargs)
     81 def forward(self, *args, **kwargs):
---> 82     return self.dynamo_ctx(self._orig_mod.forward)(*args, **kwargs)

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/eval_frame.py:209, in _TorchDynamoContext.__call__.<locals>._fn(*args, **kwargs)
    207 dynamic_ctx.__enter__()
    208 try:
--> 209     return fn(*args, **kwargs)
    210 finally:
    211     set_eval_frame(prior)

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/eval_frame.py:337, in catch_errors_wrapper.<locals>.catch_errors(frame, cache_size)
    334             return hijacked_callback(frame, cache_size, hooks)
    336 with compile_lock:
--> 337     return callback(frame, cache_size, hooks)

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py:404, in convert_frame.<locals>._convert_frame(frame, cache_size, hooks)
    402 counters["frames"]["total"] += 1
    403 try:
--> 404     result = inner_convert(frame, cache_size, hooks)
    405     counters["frames"]["ok"] += 1
    406     return result

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py:104, in wrap_convert_context.<locals>._fn(*args, **kwargs)
    102 torch.fx.graph_module._forward_from_src = fx_forward_from_src_skip_result
    103 try:
--> 104     return fn(*args, **kwargs)
    105 finally:
    106     torch._C._set_grad_enabled(prior_grad_mode)

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py:262, in convert_frame_assert.<locals>._convert_frame_assert(frame, cache_size, hooks)
    259 global initial_grad_state
    260 initial_grad_state = torch.is_grad_enabled()
--> 262 return _compile(
    263     frame.f_code,
    264     frame.f_globals,
    265     frame.f_locals,
    266     frame.f_builtins,
    267     compiler_fn,
    268     one_graph,
    269     export,
    270     hooks,
    271     frame,
    272 )

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/utils.py:163, in dynamo_timed.<locals>.dynamo_timed_inner.<locals>.time_wrapper(*args, **kwargs)
    161     compilation_metrics[key] = []
    162 t0 = time.time()
--> 163 r = func(*args, **kwargs)
    164 time_spent = time.time() - t0
    165 # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py:324, in _compile(code, globals, locals, builtins, compiler_fn, one_graph, export, hooks, frame)
    322 for attempt in itertools.count():
    323     try:
--> 324         out_code = transform_code_object(code, transform)
    325         orig_code_map[out_code] = code
    326         break

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/bytecode_transformation.py:530, in transform_code_object(code, transformations, safe)
    527 instructions = cleaned_instructions(code, safe)
    528 propagate_line_nums(instructions)
--> 530 transformations(instructions, code_options)
    531 return clean_and_assemble_instructions(instructions, keys, code_options)[1]

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/convert_frame.py:311, in _compile.<locals>.transform(instructions, code_options)
    298 nonlocal output
    299 tracer = InstructionTranslator(
    300     instructions,
    301     code,
   (...)
    309     mutated_closure_cell_contents,
    310 )
--> 311 tracer.run()
    312 output = tracer.output
    313 assert output is not None

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py:1862, in InstructionTranslator.run(self)
   1860 def run(self):
   1861     _step_logger()(logging.INFO, f"torchdynamo start tracing {self.f_code.co_name}")
-> 1862     super().run()

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py:619, in InstructionTranslatorBase.run(self)
    614 try:
    615     self.output.push_tx(self)
    616     while (
    617         self.instruction_pointer is not None
    618         and not self.output.should_exit
--> 619         and self.step()
    620     ):
    621         pass
    622 except BackendCompilerFailed:

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py:583, in InstructionTranslatorBase.step(self)
    581     if not hasattr(self, inst.opname):
    582         unimplemented(f"missing: {inst.opname}")
--> 583     getattr(self, inst.opname)(inst)
    585     return inst.opname != "RETURN_VALUE"
    586 except BackendCompilerFailed:

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/symbolic_convert.py:1941, in InstructionTranslator.RETURN_VALUE(self, inst)
   1936 _step_logger()(
   1937     logging.INFO,
   1938     f"torchdynamo done tracing {self.f_code.co_name} (RETURN_VALUE)",
   1939 )
   1940 log.debug("RETURN_VALUE triggered compile")
-> 1941 self.output.compile_subgraph(
   1942     self, reason=GraphCompileReason("return_value", [self.frame_summary()])
   1943 )
   1944 self.output.add_output_instructions([create_instruction("RETURN_VALUE")])

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py:575, in OutputGraph.compile_subgraph(self, tx, partial_convert, reason)
    572 output = []
    573 if count_calls(self.graph) != 0 or len(pass2.graph_outputs) != 0:
    574     output.extend(
--> 575         self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
    576     )
    578     if len(pass2.graph_outputs) != 0:
    579         output.append(pass2.create_store(graph_output_var))

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py:622, in OutputGraph.compile_and_call_fx_graph(self, tx, rv, root)
    620 assert_no_fake_params_or_buffers(gm)
    621 with tracing(self.tracing_context):
--> 622     compiled_fn = self.call_user_compiler(gm)
    623 compiled_fn = disable(compiled_fn)
    625 counters["stats"]["unique_graphs"] += 1

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/utils.py:163, in dynamo_timed.<locals>.dynamo_timed_inner.<locals>.time_wrapper(*args, **kwargs)
    161     compilation_metrics[key] = []
    162 t0 = time.time()
--> 163 r = func(*args, **kwargs)
    164 time_spent = time.time() - t0
    165 # print(f"Dynamo timer: key={key}, latency={latency:.2f} sec")

File /usr/local/lib/python3.9/dist-packages/torch/_dynamo/output_graph.py:709, in OutputGraph.call_user_compiler(self, gm)
    707 except Exception as e:
    708     compiled_fn = gm.forward
--> 709     raise BackendCompilerFailed(self.compiler_fn, e) from e
    710 return compiled_fn

BackendCompilerFailed: debug_wrapper raised BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.

Set torch._dynamo.config.verbose=True for more information


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True

```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0.dev20230223+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-5.4.143.bsk.8-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: A100-SXM-80GB
Nvidia driver version: 450.191.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
NUMA node(s):                    4
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8336C CPU @ 2.30GHz
Stepping:                        6
CPU MHz:                         3000.000
CPU max MHz:                     3500.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4600.00
Virtualization:                  VT-x
L1d cache:                       3 MiB
L1i cache:                       2 MiB
L2 cache:                        80 MiB
L3 cache:                        108 MiB
NUMA node0 CPU(s):               0-15,64-79
NUMA node1 CPU(s):               16-31,80-95
NUMA node2 CPU(s):               32-47,96-111
NUMA node3 CPU(s):               48-63,112-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-lightning==1.9.1
[pip3] pytorch-triton==2.0.0+d54c04abe2
[pip3] torch==2.0.0.dev20230223+cu117
[pip3] torchaudio==2.0.0.dev20230223+cu117
[pip3] torchmetrics==0.11.1
[pip3] torchvision==0.15.0.dev20230223+cu117
[conda] Could not collect
INFO[0012] Finished execution
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by weiwangmeta** [comment]
Hi @wjfwzzc, thanks for filing these issue. 

Could you please try the following version of torch? This will use a new version of triton. I have verified it works with your model. 

pip install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/test/cu117 

As of now, the above command will download 2.0.0-rc2 

Below is the output I got 

ubuntu@:~$ python nanoGPT.py                        
number of parameters: 123.69M                                                                
.../python3.10/site-packages/torch/_inductor/compile_fx.py:90: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplicati
on available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a functi
on that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issu
ecomment-1211320670 for rationale.
  warnings.warn(

Hope this helps.

**Comment by weiwangmeta** [comment]
Latest nightly also works: 
pip install torch torchvision torchAudio --index-url https://download.pytorch.org/whl/nightly/cu117
you may need to manually install jinja2 (pip install jinja2).

**Comment by bdhirsh** [comment]
I also can't repro this with my most recent pytorch + torch audio version

**Comment by wjfwzzc** [comment]
> Latest nightly also works:

That's strange, because I already used the latest nightly build (please see `Versions of relevant libraries`).

> you may need to manually install jinja2 (pip install jinja2).

And the latest jinja2 has also been installed.

Is it due to some dependencies, compile settings or environments?

**Comment by weiwangmeta** [comment]
> > Latest nightly also works:
> 
> That's strange, because I already used the latest nightly build (please see `Versions of relevant libraries`).
> 
> > you may need to manually install jinja2 (pip install jinja2).
> 
> And the latest jinja2 has also been installed.
> 
> Is it due to some dependencies, compile settings or environments?

By "the latest nightly", we really mean "the latest", like today's version: torch-2.0.0.dev20230226%2Bcu117-cp310-cp310-linux_x86_64.whl

if you uninstall your torch version and re-run this command: 
pip install torch torchvision torchAudio --index-url https://download.pytorch.org/whl/nightly/cu117 
It will install the latest nightly. 
And after you do that, your model works without error messages.   

Last week, we had resolved several issues related to triton. So 20230223 nightly must have been in broken status.

**Comment by wjfwzzc** [comment]
I tried the latest nightly just now but it failed again :(
The error is not changed.
```
Collecting environment information...
PyTorch version: 2.0.0.dev20230226+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-5.4.143.bsk.8-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: A100-SXM-80GB
Nvidia driver version: 450.191.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
NUMA node(s):                    4
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8336C CPU @ 2.30GHz
Stepping:                        6
CPU MHz:                         3000.000
CPU max MHz:                     3500.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4600.00
Virtualization:                  VT-x
L1d cache:                       3 MiB
L1i cache:                       2 MiB
L2 cache:                        80 MiB
L3 cache:                        108 MiB
NUMA node0 CPU(s):               0-15,64-79
NUMA node1 CPU(s):               16-31,80-95
NUMA node2 CPU(s):               32-47,96-111
NUMA node3 CPU(s):               48-63,112-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-lightning==1.9.1
[pip3] pytorch-triton==2.0.0+b8b470bc59
[pip3] torch==2.0.0.dev20230226+cu117
[pip3] torchaudio==2.0.0.dev20230223+cu117
[pip3] torchmetrics==0.11.1
[pip3] torchvision==0.15.0.dev20230226+cu117
[conda] Could not collect
INFO[0011] Finished execution
```

**Comment by weiwangmeta** [comment]
Can you please create a new python environment and retry? Below is the output I got. You can see my environment is very clean. 

(clean-nightly-20230226) weiwangmeta@hostname:~$ python nanoGPT.py 
number of parameters: 123.69M
/home/weiwangmeta/.conda/envs/clean-nightly-20230226/lib/python3.9/site-packages/torch/_inductor/compile_fx.py:90: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
/home/weiwangmeta/.conda/envs/clean-nightly-20230226/lib/python3.9/site-packages/torch/_functorch/aot_autograd.py:1251: UserWarning: Your compiler for AOTAutograd is returning a a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
(clean-nightly-20230226) weiwangmeta@hostname:~$ pip list |grep torch
pytorch-triton     2.0.0+b8b470bc59
torch              2.0.0.dev20230226+cu117
torchaudio         2.0.0.dev20230223+cu117
torchvision        0.15.0.dev20230226+cu117

One additional thing is: can you please try to clean the cache by "rm /tmp/torchinductor_{username}" ?

**Comment by wjfwzzc** [comment]
Your suggestion is helpful!
Use a more clean docker image instead of my company's custom image, everything works. Through step by step retries, I finally found that there was a `tensorflow==2.7.2` in the custom image and the issue would be solved if I upgraded it to `tensorflow==2.11.0`. Then I tried to install `tensorflow==2.7.2` back, found that the issue showed again.
I have no idea why the old tensorflow mysteriously hurts the latest pytorch...

**Comment by wjfwzzc** [comment]
It runs now but is much slower than `pytorch==2.0.0dev20230213`, seems due to the flash attention (`torch.backends.cuda.enable_flash_sdp(False)` recovers the speed a lot).

**Comment by ezyang** [comment]
@weiwangmeta you marked this issue as triaged but I don't see any other labels on it nor an assignment

**Comment by weiwangmeta** [comment]
@ezyang Looks like we have two issues: 
1) tensorflow and pytorch compatibility issues with tensorflow 2.7.2 
2) performance regressions if not setting torch.backends.cuda.enable_flash_sdp(False). 
I will take a look at No.1, can you please check if No.2 is expected?

**Comment by ngimel** [comment]
For 1, most likely old tensorflow links against older llvm version and doesn't hide its symbols
for 2, cc @drisspg

**Comment by drisspg** [comment]
There has only been two things effecting SDPA since 2/13: 
#95311 
#94921 
I am not entirely sure how either of these would have effected your performance

**Comment by weiwangmeta** [comment]
> Your suggestion is helpful! Use a more clean docker image instead of my company's custom image, everything works. Through step by step retries, I finally found that there was a `tensorflow==2.7.2` in the custom image and the issue would be solved if I upgraded it to `tensorflow==2.11.0`. Then I tried to install `tensorflow==2.7.2` back, found that the issue showed again. I have no idea why the old tensorflow mysteriously hurts the latest pytorch...

Thanks @wjfwzzc 
It seems the issue might be setup specific. I installed tensorflow==2.7.2 (and later tensorflow-gpu==2.7.2), I could not reproduce this issue. 

(clean-nightly-20230226) weiwangmeta@hostname:~$ python nanoGPT.py  
number of parameters: 123.69M
/home/weiwangmeta/.conda/envs/clean-nightly-20230226/lib/python3.9/site-packages/torch/_inductor/compile_fx.py:90: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
(clean-nightly-20230226) weiwangmeta@hostname:~$ pip list |grep tensorflow
tensorflow-estimator         2.7.0
tensorflow-gpu               2.7.2
tensorflow-io-gcs-filesystem 0.31.0
(clean-nightly-20230226) weiwangmeta@hostname:~$ pip list |grep torch
pytorch-triton               2.0.0+b8b470bc59
torch                        2.0.0.dev20230227+cu117
torchaudio                   2.0.0.dev20230223+cu117
torchvision                  0.15.0.dev20230227+cu117

**Comment by ngimel** [comment]
@drisspg can you reproduce perf difference with torch.backends.cuda.enable_flash_sdp(False)? Is enabling flash expected to regress perf?

**Comment by drisspg** [comment]
I just pulled nanogpt master.
torch.compile + flashattention : 3858.03ms
torch.compile + mem_efficient_attention: 4112.60ms

**Comment by wjfwzzc** [comment]
I'll close this issue because I cannot reproduce it on pytorch 2.0 stable version.
